src/main/java/org/elasticsearch/common/geo/builders/BasePolygonBuilder.java

```diff
@@ -261,20 +278,54 @@ private static int component(final Edge edge, final int id, final ArrayList<Edge
         // run along the border of the component, collect the
         // edges, shift them according to the dateline and
         // update the component id
-        int length = 0;
+        int length = 0, connectedComponents = 0;
+        // if there are two connected components, splitIndex keeps track of where to split the edge array
+        // start at 1 since the source coordinate is shared
+        int splitIndex = 1;
         Edge current = edge;
+        Edge prev = edge;
+        // bookkeep the source and sink of each visited coordinate
+        WeakHashMap<Coordinate, Pair<Edge, Edge>> visitedEdge = new WeakHashMap<>();
         do {
-
-            current.coordinate = shift(current.coordinate, shift); 
+            current.coordinate = shift(current.coordinate, shift);
```

Comment:
I know this was already here, but could you rename the variable shift to something that doesn't collide with the function name? Maybe datelineShift?